{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This script processes the CDHit files and generates a csv with peptides and read counts for each peptide.\n",
    "#Works around CDHit's issue of selecting a sequence randomly as the representative sequence.\n",
    "#Use CDHit file which now has the sequence as part of the name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import csv\n",
    "import string\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import filecmp\n",
    "from scipy import stats\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_2_aa = {\"TTT\":\"F\", \"TTC\":\"F\", \"TTA\":\"L\", \"TTG\":\"L\",\n",
    "    \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\",\n",
    "    \"TAT\":\"Y\", \"TAC\":\"Y\", \"TAA\":\"*\", \"TAG\":\"*\",\n",
    "    \"TGT\":\"C\", \"TGC\":\"C\", \"TGG\":\"W\", \"TGA\":\"*\",\n",
    "    \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "    \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "    \"CAT\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "    \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "    \"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"I\", \"ATG\":\"M\",\n",
    "    \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "    \"AAT\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "    \"AGT\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "    \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "    \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "    \"GAT\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "    \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\"}\n",
    "\n",
    "def translate(pep_dna):\n",
    "#Translate peptide to amino acids\n",
    "    codons_aa = []\n",
    "    codons_dna = []\n",
    "    for j in range(0,len(pep_dna),3):   \n",
    "        codons_dna.append(pep_dna[j:j+3])\n",
    "    codons_aa = [codon_2_aa[x] for x in codons_dna] # translate\n",
    "    pep_aa=''.join(codons_aa) #puts letters into string\n",
    "    return pep_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bad_clusters_fn(file_name):\n",
    "#Identify which cluster numbers have incorrectly selected a representative sequence.\n",
    "#goes through .clstr file\n",
    "#calculates the percent of members of each cluster with sequence identity at 100% or other percent\n",
    "#outputs the cluster number and the number of total peptides and the number of peptides that match the representative sequence\n",
    "    total_list=[]\n",
    "    matches_list=[]\n",
    "    cluster_list=[]\n",
    "\n",
    "    matches = 0\n",
    "    total = 0\n",
    "\n",
    "    with open(file_name, 'r') as read_file:  \n",
    "        for line in read_file:\n",
    "            if line[0] == '>':\n",
    "                cluster = line[line.strip().find('Cluster ')+len('Cluster '):-1]\n",
    "                cluster_list.append(cluster)\n",
    "                matches_list.append(matches+1)#add 1 because the rep seq is at 100%\n",
    "                total_list.append(total)\n",
    "                matches = 0\n",
    "                total = 0\n",
    "            else:\n",
    "                total+=1\n",
    "                if '%' in line:\n",
    "                    percent = line[line.find('+/')+len('+/'):line.rfind('%')]\n",
    "                    if percent == '100.00':\n",
    "                        matches+=1\n",
    "        matches_list.append(matches+1)#add 1 because the rep seq is at 100%\n",
    "        total_list.append(total)             \n",
    "\n",
    "    clusters = pd.DataFrame({\n",
    "        'cluster#':cluster_list,\n",
    "        'perfect_match':matches_list[1:],\n",
    "        'total_seq':total_list[1:]})\n",
    "\n",
    "    clusters['frac_100'] = clusters['perfect_match']/clusters['total_seq']\n",
    "\n",
    "    #select subset of clusters with fraction of members with perfect sequence identity to representative sequence below 50%\n",
    "    bad_clusters = clusters[ clusters['frac_100']< 0.50 ]   \n",
    "    print('Number of clusters with incorrect representative sequence assigned by CDHit:',len(bad_clusters))\n",
    "    #bad_clusters.to_csv(roundx+'_badclusters.csv', index=False)\n",
    "    return bad_clusters\n",
    "\n",
    "def bad_cluster_members_fn(file_name, bad_clusters):\n",
    "# Go through CDhit file and pull out the sequence names corresponding to clusters \n",
    "## that have a representative sequence which doesn't match well to the other sequences\n",
    "    cluster_list = []\n",
    "    sequence_name_list = []\n",
    "    seq_list = []\n",
    "    percent_list = []\n",
    "    the_clust_flag = 0\n",
    "\n",
    "    i=0\n",
    "    with open(file_name, 'r') as read_file:  \n",
    "        for line in read_file:\n",
    "            #check if this cluster number is in the list of 'offending' clusters\n",
    "            if line[0]=='>':\n",
    "                line_cluster = line.strip()[line.strip().find('Cluster ')+len('Cluster '):] \n",
    "                if i<len(bad_clusters):\n",
    "                    if line_cluster==bad_clusters['cluster#'].iloc[i]:\n",
    "                        the_clust_flag = 1\n",
    "                        i+=1\n",
    "                    else:\n",
    "                        the_clust_flag = 0\n",
    "                else:\n",
    "                    the_clust_flag = 0\n",
    "            if line[0]!='>' and the_clust_flag==1:\n",
    "                cluster_list.append(line_cluster)\n",
    "                \n",
    "                sequence_name=line[line.find('39nt, ')+len('39nt, '):line.find('...')]\n",
    "                sequence_name_list.append(sequence_name.strip())\n",
    "                seq_list.append(sequence_name.strip().split('_')[1])\n",
    "\n",
    "                percent = line.strip()[line.strip().find('+/')+len('+/'):line.strip().find('%')]\n",
    "                if line.strip()[-1]=='*':\n",
    "                    percent=float(100)\n",
    "                percent_list.append(float(percent))\n",
    "\n",
    "            if i>len(bad_clusters):\n",
    "                break\n",
    "    \n",
    "    bugged = pd.DataFrame({\n",
    "        'cluster_num':cluster_list,\n",
    "        'seq_id':sequence_name_list,\n",
    "        'percent_id':percent_list,\n",
    "        'dna':seq_list\n",
    "        })\n",
    "\n",
    "    bad_cluster_members = bugged\n",
    "    return bad_cluster_members\n",
    "\n",
    "\n",
    "def count_fn(bad_cluster_members):\n",
    "#count number of times that DNA showed up in a cluster and the number of sequences in the cluster\n",
    "    #count # of times that DNA seq showed up in bad_cluster_members\n",
    "    dna_count = bad_cluster_members.groupby('dna')['seq_id'].nunique()\n",
    "    dna_count_df = pd.DataFrame({'dna':dna_count.index, 'count_dna':dna_count.values}) #makes into a dataframe\n",
    "    bugged_dna_count = pd.merge(dna_count_df, bad_cluster_members, on=['dna']) #was df_counts\n",
    "\n",
    "    #count # of sequences in that cluster\n",
    "    cluster_count = bugged_dna_count.groupby('cluster_num')['seq_id'].nunique()\n",
    "    cluster_count_df = pd.DataFrame({'cluster_num':cluster_count.index, 'count_clust':cluster_count.values}) #makes into a dataframe\n",
    "\n",
    "    bad_cluster_members_count = pd.merge(cluster_count_df, bugged_dna_count, on=['cluster_num'])\n",
    "    bad_cluster_members_count['frac'] = bad_cluster_members_count['count_dna']/bad_cluster_members_count['count_clust']\n",
    "    return bad_cluster_members_count\n",
    "\n",
    "def replace_list_fn(bad_cluster_members_count):\n",
    "    old_list=[]\n",
    "    new_list=[]\n",
    "\n",
    "    for x in bad_cluster_members_count['cluster_num'].unique():\n",
    "        y = bad_cluster_members_count[bad_cluster_members_count['cluster_num'] == x]\n",
    "        old = y['dna'].loc[y['percent_id'].idxmax()] #old sequence has 100% identity to representative seq (is the old rep seq)\n",
    "        old_list.append(old)\n",
    "        new = y['dna'].loc[y['count_dna'].idxmax()] #take sequence with max number of sequences like it\n",
    "        new_list.append(new)\n",
    "\n",
    "    replacement_df = pd.DataFrame({'old_rep':old_list, 'new_rep':new_list})\n",
    "    \n",
    "    return replacement_df\n",
    "\n",
    "\n",
    "#calls other functions for debugging CDHit\n",
    "def fix_cdhit_bug(file_name):\n",
    "    file_name_clstr = file_name + '.clstr'\n",
    "    bad_clusters = bad_clusters_fn(file_name_clstr)\n",
    "    bad_cluster_members = bad_cluster_members_fn(file_name_clstr, bad_clusters) \n",
    "    bad_cluster_members_count = count_fn(bad_cluster_members)\n",
    "    replacement_df = replace_list_fn(bad_cluster_members_count)\n",
    "    return replacement_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_cdhit_rep(file_name, replacement_df):\n",
    "#Import sequences from CD-Hit representative-sequences file\n",
    "##In sets of twos, lines contain [Name, Sequence].  Import every second line (not the name line)\n",
    "    read_list = []\n",
    "    with open(file_name, 'r') as read_file:  \n",
    "        for line in read_file:\n",
    "            if '>' not in line:\n",
    "                #check if line is in offenders list for wrong representative sequence chosen by CDHit\n",
    "                if line.strip() in replacement_df['old_rep'].values:\n",
    "                    new = replacement_df[replacement_df['old_rep'].isin([line.strip()])]['new_rep'].values\n",
    "                    read_list.append(translate(new.item()))\n",
    "                else:\n",
    "                    read_list.append(translate(line.strip('\\n')))\n",
    "                    \n",
    "    return read_list\n",
    "\n",
    "def import_cdhit_clstr(file_name):\n",
    "#Import sequences from CD-Hit cluster file\n",
    "    #\n",
    "    ##The file contains a list of clusters of the format:\n",
    "    ### >Cluster X\n",
    "    ### 0       39nt, >1101:14633:1491:ATC... *\n",
    "    ### 1       39nt, >1101:12909:1680:ATC... at +/100.00%\n",
    "    ### etc\n",
    "    #Look for the '>' flag for the new cluster then extract the number of sequences in the previous cluster\n",
    "    #\n",
    "    clstr_file_name = file_name + '.clstr' #file containing clusters\n",
    "    cluster_count = 0\n",
    "    first_line_flag = 1\n",
    "    count_list = [] \n",
    "    with open(clstr_file_name, 'r') as read_file:  \n",
    "        for line in read_file:\n",
    "            #the line preceding a new cluster tells the number of sequences in that previous cluster\n",
    "            if first_line_flag==1:\n",
    "                first_line_flag=0\n",
    "                line_before=line\n",
    "            else:\n",
    "                if line[0] == '>': \n",
    "                    count_list.append( int(line_before.split('\\t')[0])+1 ) #start at 0, so add 1 to cluster number label\n",
    "\n",
    "                    cluster_count=cluster_count+1\n",
    "                line_before=line\n",
    "        #Last line\n",
    "        count_list.append( int(line_before.split('\\t')[0])+1 )\n",
    "        cluster_count=cluster_count+1\n",
    "        #print(\"Number of unique DNA sequences (with >1 nucleotide difference):\", cluster_count)\n",
    "    return count_list\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def confirm_order(rep_file_name):\n",
    "# Confirm that the representative sequences in order match the representative sequences of the clusters\n",
    "# CD-Hit documentation does not explicitly confirm\n",
    "    with open(rep_file_name+'.clstr', 'r') as read_file, open('rep_check', 'w', newline='') as write_file:\n",
    "        for line in read_file:\n",
    "            if '*' in line:\n",
    "                pep=line.split('>')[-1].rsplit(':',1)[0] #take sequence between the '>' and the last ':'\n",
    "                write_file.write(pep+ '\\n')\n",
    "    with open(rep_file_name, 'r') as read_file, open('clstr_check', 'w', newline='') as write_file:\n",
    "        for line in read_file:\n",
    "            if '>' in line:\n",
    "                pep=line.split('>')[-1].rsplit(':',1)[0] \n",
    "                write_file.write(pep+ '\\n')\n",
    "    if filecmp.cmp('rep_check', 'clstr_check')==False: # filecmp returns true if they seem equal\n",
    "        print('ERROR: Representative sequence order incorrect')\n",
    "    else:\n",
    "        print('Proceed: Representative sequence order correct')\n",
    "        \n",
    "def collapse2uniqAA(illumina):\n",
    "#Some DNA sequences are unique but, when translated, are non-unique AA sequences. Combine non-unique AA sequences(e.g. ALPWRSVWL)\n",
    "    illumina_nodup=illumina.groupby(['sequence']).sum().reset_index()  \n",
    "    if np.sum(illumina_nodup['count'])==np.sum(illumina['count']):\n",
    "        print('Proceed: Number of reads unchanged while collapsing AA sequence duplicates')\n",
    "    else:\n",
    "        print('Error: Number of reads changed while collapsing AA duplicates')\n",
    "    return illumina_nodup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DR401_R1\n",
      "Number of clusters with incorrect representative sequence assigned by CDHit: 116\n",
      "Proceed: Representative sequence order correct\n",
      "Proceed: Number of reads unchanged while collapsing AA sequence duplicates\n",
      "DR401_R2\n",
      "Number of clusters with incorrect representative sequence assigned by CDHit: 193\n",
      "Proceed: Representative sequence order correct\n",
      "Proceed: Number of reads unchanged while collapsing AA sequence duplicates\n",
      "DR401_R3\n",
      "Number of clusters with incorrect representative sequence assigned by CDHit: 252\n",
      "Proceed: Representative sequence order correct\n",
      "Proceed: Number of reads unchanged while collapsing AA sequence duplicates\n"
     ]
    }
   ],
   "source": [
    "def import_cdhit_all( roundx):\n",
    "    print( roundx)\n",
    "    rep_file_name = 'pepdist_{}_cdhit'.format(roundx)\n",
    "    \n",
    "    replacement_df = fix_cdhit_bug(rep_file_name)\n",
    "    read_list = import_cdhit_rep(rep_file_name, replacement_df)\n",
    "    count_list = import_cdhit_clstr(rep_file_name)\n",
    "    confirm_order(rep_file_name)\n",
    "    \n",
    "    #Combine sequences and read count for each sequence\n",
    "    illumina = pd.DataFrame(\n",
    "    {'sequence': read_list,\n",
    "     #'dna': dna_list,\n",
    "     'count': count_list,\n",
    "    })\n",
    "    \n",
    "    illumina_nodup = collapse2uniqAA(illumina)\n",
    "    \n",
    "    return illumina_nodup\n",
    "\n",
    "\n",
    "for roundx in [\"DR401_R1\",\"DR401_R2\",\"DR401_R3\"]:\n",
    "    illumina_nodup = import_cdhit_all( roundx)\n",
    "    illumina_nodup.to_csv('{}_cdhit-corrected_data.csv'.format(roundx))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
